src/Servers/Kestrel/Core/src/Internal/Http2/Http2StreamStack.cs

```diff
@@ -78,7 +79,59 @@ private void PushWithResize(Http2Stream item)
             _size++;
         }
 
-        private readonly struct Http2StreamAsValueType
+        public void RemoveExpired(long now)
+        {
+            int size = _size;
+            Http2StreamAsValueType[] array = _array;
+
+            var removeCount = CalculateRemoveCount(now, size, array);
+            if (removeCount == 0)
+            {
+                return;
+            }
+
+            var newSize = size - removeCount;
+
+            // Dispose removed streams
+            for (var i = 0; i < removeCount; i++)
+            {
+                Http2Stream stream = array[i];
+                stream.Dispose();
+            }
+
+            // Move remaining streams
+            for (var i = 0; i < newSize; i++)
+            {
+                array[i] = array[i + removeCount];
+            }
+
+            // Clear unused array indexes
+            for (var i = newSize; i < size; i++)
+            {
+                array[i] = default;
+            }
+
+            _size = size - removeCount;
```

Comment:
Nit:

```suggestion
            _size = newSize;
```

Any reason you can't set `_size` to `newSize` as soon as it's calculated above? `size` is stored in a locally anyway.

Reply:
I added newSize later and forgot to update this 😄